design-proposals/vm-pool.md

```diff
@@ -0,0 +1,194 @@
+* As a SRE managing the availability of VM fleets I want to automate VM recovery by auto detecting and deleting misbehaving VMs and having the platform spin up fresh new instances as a replacement.
+* As a pool user/manager I want to remove a VM from the pool without modifying it for debugging. The missing VM can be replaced by the pool.
+
+# Design
+
+The VMPool design introduces two new APIs represented as CRDs, the VirtualMachinePool (VMPool) object and the VirtualMachineConfig (VMConfig) object. For the most part, a VMConfig object is simply a VM object without a status section that is used by a VMPool to stamp out multiple replicas.
+
+The figure below illustrates the relationship between a VMPool, VMConfig, and VM. The VMPool stamps out new VM replicas using the contents as in the VMConfig.
+
+
+![VMPool+VMConfig+VM](vm-pool-images/figure1.png "VMPool and VMConfig relationship")
+
+In order to make this even more tangible, below is a figure that outlines the types of tunings present in each of these APIs.
+
+![VMPool vs. VMConfig](vm-pool-images/figure2.png "VMPool vs. VMConfig")
+
+## VirtualMachineConfig (VMConfig) API
+
+The VMConfig api contains a VirtualMachineSpec structure that is identical to a VM object, which makes it similar to a VM object itself. However there are some key differences here between the two objects.
+
+**VMConfig vs VM**
+
+**Differences**
+* A VMConfig object can never be started. Instead the VMConfig object is referenced by higher level controllers (like a VMPool) which can start many instances of VMs using the VMConfig.
+* The VMConfig has a different status section from VM
+* A VMConfig’s name does not correlate in any way to a VM generated from the VMConfig
+* In a VMConfig, all naming resulting in newly created objects are treated as generated names in practice. For example, the name of a DataVolumeTemplate in a VMConfig.Spec.DataVolumeTemplates list will be treated as a generated name when VMs are spawned using the config. This ensures all VMs derived from a VMConfig are unique.
+
+**Similarities**
+* VMConfig and VM object’s spec structure are identical
+
+## VirtualMachinePool (VMPool) API
+
+The VMPool api represents all the tunings necessary for managing a pool of stateful VMs. This API is very simple and leverages the VMConfig object for most of the heavy lifting pertaining to what the VM instances within the pool actually look like.
+
+The VMPools spec contains the following tunings and values
+
+* **VirtualMachineRef** - (Required) A structure that points to a VMConfig object
+* **Replicas** - (Required) An integer representing the desired number of VM replicas
+* **MaxUnavailable**  - (Optional) (Defaults to 25%) Integer or string pointer, that when set represents either a percentage or number of VMs in a pool that can be unavailable (ready condition false) at a time during automated update.
```

Comment:
Here we could borrow from https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/stateful-set-v1/#StatefulSetSpec to be consistent and keep our options open.
So move anything regarding updates into a `updateStrategy` field as well instead of having it top-level.
That also matches the fact that the design already contains a separate `disruptionStrategy`.

Reply:
the disruption strategy was meant to combine how to handle both updates and scale-in. Are you saying you think those should be treated separately?